docs/Contributing/Fleet-UI-Testing.md

```diff
@@ -0,0 +1,362 @@
+    await user.click(screen.getByRole("button", { name: "Reset password" }));
+
+    const passwordError = screen.getByText(
+      "Password must meet the criteria below"
+    );
+    expect(passwordError).toBeInTheDocument();
+    expect(submitSpy).not.toHaveBeenCalled();
+  });
+});
+```
+
+#### E2E testing
+
+Our e2e layer tests all the systems of the software (frontend and backend) together to
+ensure the application works as intended. To support this we do not often mock API responses at this
+layer of testing (with an exception for mocking network error responses). We want to test the
+software as an actual user would at this level.
+
+[source here](https://github.com/fleetdm/fleet/blob/main/cypress/integration/all/app/labelflow.spec.ts)
+
+```ts
+describe("Labels flow", () => {
+  before(() => {
+    // ...setup
+  });
+  after(() => {
+    // ...teardown
+  });
+
+  describe("Manage hosts page", () => {
+    beforeEach(() => {
+      // ...setup
+    });
+    it("creates a custom label", () => {
+      cy.getAttached(".label-filter-select__control").click();
+      cy.findByRole("button", { name: /add label/i }).click();
+      cy.getAttached(".ace_content").type(
+        "{selectall}{backspace}SELECT * FROM users;"
+      );
+      cy.findByLabelText(/name/i).click().type("Show all MAC users");
+      cy.findByLabelText(/description/i)
+        .click()
+        .type("Select all MAC users.");
+      cy.getAttached(".label-form__form-field--platform > .Select").click();
+      cy.getAttached(".Select-menu-outer").within(() => {
+        cy.findByText(/macOS/i).click();
+      });
+      cy.findByRole("button", { name: /save label/i }).click();
+      cy.findByText(/label created/i).should("exist");
+    });
+  });
+});
+```
+
+
+### Tooling
+
+Here is a quick reference of the current tooling we are using at each layer of testing.
+
+<img src="https://miro.medium.com/max/1400/1*iBBcTAf4zvn7yZq4K4MShA.png" width="400">
+
+#### Eslint and Typescript
+
+We use these for our static analyisis testing. The linting and typing rules have been setup so
+errors should appear in your editor if they are broken.
+
+#### Jest
+
+We use jest our our frontend test runner, assertion library, and spy and mock utilities for unit and
+integration testing.
+
+#### Testing-Library
+
+We rely heavily on the different libraries that are part of the testing-library ecosystem for out
+unit and integration testing. These including react-testing-library, cypress-testing-library,
+react-hooks, and user-events. The guiding principles of testing-libary align with our own in that we
+believe tests should resemble real world usage as closely as possible.
+
+#### Cypress
+
+We use cypress with cypress-testing-library as our e2e testing framework. We primarialy rely on full e2e software testing and do
+not often mock API responses, but we make an exception for mocking a failed network response when
+testing error states of the application.
```

Comment:
Add:
and testing external APIs

Reply:
Add:

For more details on our E2E testing, check out our [Cypress documentation](../cypress/README.md).